### PR TITLE
New operators linalg_syrk, linalg_gelqf

### PR DESCRIPTION
**Pull request by mseeger**
- Works for CPU only for now. We will supply the GPU versions next
- Added dtype parameters to unit test support code, so we can run numerical tests in float64

@asmushetzel

### CODE REVIEW

File: src/operator/tensor/la_op.cc [hunk 1]
```diff
@@ -95,33 +96,33 @@ NNVM_REGISTER_OP(_backward_linalg_gemm)
...
-.. note:: The operator does only support float32 and float64 data types and provides
-          proper backward gradients.
+.. note:: The operator supports float32 and float64 data types only.
 
 Examples::
 
    // Single matrix multiply
    A = [[1.0, 1.0], [1.0, 1.0]]
    B = [[1.0, 1.0], [1.0, 1.0], [1.0, 1.0]]
-   gemm2(A, B, transpose_b = 1, alpha = 2.0)
```
Comment (asmushetzel) [inline src/operator/tensor/la_op.cc]: Apparently some general renaming has happened recently for namespace reasons. All documentation now uses function names without linalg_prefix.

File: src/operator/tensor/la_op.cc [hunk 2]
```diff
@@ -251,42 +253,42 @@ NNVM_REGISTER_OP(_backward_linalg_potri)
 .set_attr<nnvm::TIsBackward>("TIsBackward", true)
 .set_attr<FCompute>("FCompute<cpu>", LaOpBackward<cpu, 2, 2, 3, 1, potri_backward>);
 
-NNVM_REGISTER_OP(_linalg_trmm)
+NNVM_REGISTER_OP(linalg_trmm)
```
Comment (asmushetzel) [inline src/operator/tensor/la_op.cc]: All these operators are now registered with an initial underscore followed by an alias for backward compatibility.

File: src/operator/tensor/la_op.cc [hunk 3]
```diff
@@ -420,5 +421,134 @@ NNVM_REGISTER_OP(_backward_linalg_sumlogdiag)
 .set_attr<nnvm::TIsBackward>("TIsBackward", true)
 .set_attr<FCompute>("FCompute<cpu>", LaOpBackward<cpu, 2, 2, 2, 1, sumlogdiag_backward>);
 
+NNVM_REGISTER_OP(linalg_syrk)
```
Comment (asmushetzel) [inline src/operator/tensor/la_op.cc]: Initial underscore followed by alias (as well as in the other new operators)

File: src/operator/tensor/la_op.h [hunk 4]
```diff
@@ -200,9 +216,7 @@ inline bool LaReduceShape(const nnvm::NodeAttrs& attrs,
   CHECK_EQ(in_attrs->size(), 1);
   CHECK_EQ(out_attrs->size(), 1);
   const int ndim((*in_attrs)[0].ndim());
-  if ( ndim < dim ) {
-     return false;
-  }
+  CHECK_GE(ndim, dim) << "Shape of input has too few dimensions";
```
Comment (asmushetzel) [inline src/operator/tensor/la_op.h]: Please leave original code. We may have ndim < dim in the case where the input shape has not yet been defined at all (but may be defined later). Means at this point in time we can't do any inference but not necessary that we can't ever (in a subsequent pass).

File: src/operator/tensor/la_op.cc (hunk 1 above)
Comment (mseeger) [reply src/operator/tensor/la_op.cc]: OK, I will change the docstrings

File: src/operator/tensor/la_op.cc (hunk 2 above)
Comment (mseeger) [reply src/operator/tensor/la_op.cc]: OK, I will do this

File: src/operator/tensor/la_op.cc (hunk 3 above)
Comment (mseeger) [reply src/operator/tensor/la_op.cc]: OK, will do this

File: src/operator/linalg.h [hunk 5]
```diff
@@ -123,6 +123,48 @@ void linalg_potri(const Tensor<xpu, 2, DType>& A, bool lower, Stream<xpu> *s = 0
...
+void linalg_batch_syrk(const Tensor<xpu, 3, DType>& A,
+                       const Tensor<xpu, 3, DType>& B, DType alpha, DType beta,
+                       bool tA, Stream<xpu> *s = 0);
+
+//////////////////////////////// GELQF ////////////////////////////////////////////
+
+// CPU/GPU-versions of LAPACK functions "gelqf", "orglq". Please refer to the
+// LAPACK documentation for further details.
+// Note:
+// - The current implementation works for CPU only. In particular, when called
```
Comment (asmushetzel) [inline src/operator/linalg.h]: There are no  separate batch mode functions for this. So we should not refer to batch mode functions in this comment or add batch mode functions.

File: docs/api/python/symbol.md [hunk 6]
```diff
@@ -526,6 +526,8 @@ Composite multiple symbols into a new one by an operator.
     linalg_trmm
     linalg_trsm
     linalg_sumlogdiag
+	linalg_syrk
```
Comment (piiswrong) [inline docs/api/python/symbol.md]: wrong indentation

**Comment by piiswrong** [comment]
what's the status on this?

File: docs/api/python/symbol.md (hunk 6 above)
Comment (mseeger) [reply docs/api/python/symbol.md]: Done

**Comment by mseeger** [comment]
I managed to run all unit tests in continuous-integration/jenkins/pr-head, except for the R ones. They fail for funny reasons, which seem nothing to do with my changes.

I am trying again.

**Comment by mseeger** [comment]
Unit tests for esoteric APIs (Perl, R, Scala) fail with:

Remote call on mxnet14 failed

I fail to see what this has to do with my code changes. Not interested in these APIs (or should I be?).

**Comment by asmushetzel** [comment]
The R/Perl-unit tests were failing because of a git-issue. All other ones did pass. I did a full review and IMO all is fine now.  

There will be a followup PR by me in a week or so that will bring in the CUDA support for these new operators. 

So from my point of view, this can be integrated. Just would like to have confirmation from Matthias that the documentation changes look as expected when formatted.

**Comment by mseeger** [comment]
Yes I confirmed all docstrings in the tool you sent me, they all look fine.

**Comment by asmushetzel** [comment]
Thank you Eric for the incredible turnaround time

File: docs/api/python/symbol.md [hunk 7]
```diff
@@ -526,6 +526,8 @@ Composite multiple symbols into a new one by an operator.
     linalg_trmm
     linalg_trsm
     linalg_sumlogdiag
+    linalg_syrk
+    linalg_gelqf
```
Comment (eric-haibin-lin) [inline docs/api/python/symbol.md]: I assume `ndarray.md` should be updated, too?

File: src/operator/linalg_impl.h [hunk 8]
```diff
@@ -719,4 +719,137 @@ LINALG_GPU_BATCH_POTRI(double)
...
+    << "Non compatible matrix dimensions between inputs A and B for syrk";
+}
+
+#define LINALG_CPU_SYRK(fname, DType) \
+template<> inline \
+void linalg_syrk<cpu, DType>(const Tensor<cpu, 2, DType>& A, \
+                             const Tensor<cpu, 2, DType>& B, DType alpha, \
+                             DType beta, bool tA, Stream<cpu> *s) { \
+  check_syrk(A, B, alpha, beta, tA); \
+  cblas_##fname(CblasRowMajor, CblasLower, (tA ? CblasTrans : CblasNoTrans), \
```
[the remark this answers is not in the document]
Comment (mseeger) [reply src/operator/linalg_impl.h]: Hello, {s|d}syrk is a BLAS function, which can be called as cblas_{s|d}syrk (via the cblas interface).
For whatever reason, this does not work in your particular setup. What is quite odd about this is that linalg_impl.h calls a range of other cblas_XXX functions:
cblas_*gemm, cblas_*trmm, cblas_*trsm (where * in s, d).

AFAIK, cblas has all of these AND cblas_*syrk.

Are you getting errors for the other cblas calls as well?

Maybe @asmushetzel has an idea what is going on here?

[the remark this answers is not in the document]
Comment (iblis17) [reply src/operator/linalg_impl.h]: I finally figure out that MXNet.jl need a modified `cblas.h`: https://github.com/dmlc/MXNet.jl/blob/master/deps/cblas.h

I will manage to send a patch for that. @mseeger thanks for your time and explanation!

[the remark this answers is not in the document]
Comment (mseeger) [reply src/operator/linalg_impl.h]: It seems that your cblas.h is edited, right? This is not good. Maybe find out what Julia is doing there?
We may rely on further blas functions in the future, so better try and get the correct cblas.h being used

**Comment by mseeger** [comment]
This was also my thought. Why do they edit cblas.h? Just to save some compile time?

**Pull request merged**